sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -949,14 +949,19 @@ class Analyzer(
             if oldVersion.outputSet.intersect(conflictingAttributes).nonEmpty =>
           (oldVersion, oldVersion.copy(serializer = oldVersion.serializer.map(_.newInstance())))
 
-        // Handle projects that create conflicting aliases.
         case oldVersion @ Project(projectList, _)
-            if findAliases(projectList).intersect(conflictingAttributes).nonEmpty =>
-          (oldVersion, oldVersion.copy(projectList = newAliases(projectList)))
+            if hasConflict(projectList, conflictingAttributes) =>
+          (oldVersion,
+            oldVersion.copy(
+              projectList =
+                newNamedExpression(projectList, conflictingAttributes)))
 
         case oldVersion @ Aggregate(_, aggregateExpressions, _)
```

Comment:
It seems that the query in the PR description works well with the fix below, too;
```
        case oldVersion @ Aggregate(_, aggregateExpressions, _)
            if AttributeSet(aggregateExpressions.map(_.toAttribute))
              .intersect(conflictingAttributes).nonEmpty =>
        (oldVersion, oldVersion.copy(aggregateExpressions = aggregateExpressions.map(_.newInstance())))
```
Could we fix this issue in an easier way than the current fix?

Reply:
> Could we fix this issue in an easier way than the current fix?

I don't think it is robust enough. For example, the following test fails with the suggested fix:
```
[info] - [SPARK-6231] join - self join auto resolve ambiguity *** FAILED *** (251 milliseconds)
[info]   Failed to analyze query: org.apache.spark.sql.AnalysisException: Resolved attribute(s) key#4619 missing from key#4518,value#4519 in operator !Aggregate [key#4619], [key#4619, sum(cast(key#4619 as bigint)) AS sum(key)#4620L]. Attribute(s) with the same name appear in the operation: key. Please check if the right attribute(s) are used.;;
[info]   Join Inner, (key#4518 = key#4518)
[info]   :- Aggregate [key#4518], [key#4518, count(1) AS count(1)#4610L]
[info]   :  +- Project [_1#4513 AS key#4518, _2#4514 AS value#4519]
[info]   :     +- LocalRelation [_1#4513, _2#4514]
[info]   +- !Aggregate [key#4619], [key#4619, sum(cast(key#4619 as bigint)) AS sum(key)#4620L]
[info]      +- Project [_1#4513 AS key#4518, _2#4514 AS value#4519]
[info]         +- LocalRelation [_1#4513, _2#4514]
[info]   
```

Reply:
Ur, I see... In the query you showed in the PR description, it seems the dedup logic doesn't work in the `Expand` node (^^^^^ below):
```
'Join Inner
:- Aggregate [nums#121, spark_grouping_id#119], [nums#121, max(0) AS agcol#118]
:  +- Expand [List(nums#79, nums#120, 0), List(nums#79, null, 1)], [nums#79, nums#121, spark_grouping_id#119]
:     +- Project [nums#79, nums#79 AS nums#120]
:        +- Project [value#76 AS nums#79]
:           +- LocalRelation [value#76]
+- Aggregate [nums#121, spark_grouping_id#119], [nums#121, max(0) AS agcol#124]
              ^^^^^^^^
   +- Expand [List(nums#79, nums#120, 0), List(nums#79, null, 1)], [nums#79, nums#121, spark_grouping_id#119]
                                                                             ^^^^^^^^
      +- Project [nums#79, nums#79 AS nums#120]
         +- Project [value#76 AS nums#79]
            +- LocalRelation [value#76]
```
So, we might be able to fix this dedup issue by adding an entry for `Expand` in `dedupRight` like this?;
```
        case oldVersion @ Expand(_, output, _)
            if oldVersion.outputSet.intersect(conflictingAttributes).nonEmpty =>
          (oldVersion, oldVersion.copy(output = output.map(_.newInstance())))
```